File: frontend/src/api/splitting.ts
import { axiosInstance as apiClient } from './tools/instance';

export interface SplitOptions {
  bill_id: number;
  subtotal: number;
  tax_amount: number;
  service_fee_amount: number;
  total_amount: number;
  items: Array<{
    id: string;
    name: string;
    price: number;
    quantity: number;
    subtotal: number;
  }>;
  max_people: number;
}

export interface SplitResult {
  bill_id: number;
  split_method: 'equal' | 'custom' | 'items';
  total_amount: number;
  tip_amount: number;
  grand_total: number;
  people: Array<{
    person_id: string;
    name: string;
    base_amount: number;
    tax_amount: number;
    service_fee_amount: number;
    tip_amount: number;
    total_amount: number;
    items?: Array<{
      id: string;
      name: string;
      price: number;
      quantity: number;
      subtotal: number;
    }>;
  }>;
  created_at: string;
}

export interface EqualSplitRequest {
  bill_id: number;
  num_people: number;
  people: Record<string, string>; // person_id -> name
}

export interface CustomSplitRequest {
  bill_id: number;
  amounts: Record<string, number>; // person_id -> amount
  people: Record<string, string>; // person_id -> name
}

export interface ItemSplitRequest {
  bill_id: number;
  item_selections: Record<string, string[]>; // person_id -> item_ids
  people: Record<string, string>; // person_id -> name
}

export interface SplitValidationRequest {
  bill_id: number;
  split_method: 'equal' | 'custom' | 'items';
  data: EqualSplitRequest | CustomSplitRequest | ItemSplitRequest;
}

export interface SplitValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
  total_check: {
    expected: number;
    calculated: number;
    difference: number;
  };
}

export class SplittingAPI {
  /**
   * Get split options for a bill
   */
  static async getSplitOptions(billId: number): Promise<SplitOptions> {
    const response = await apiClient.get(`/bills/${billId}/split/options`);
    return response.data.result;
  }

  // Execute split payment coordination
  static async executeSplitPayment(billId: number, splitResult: SplitResult, paymentInfo: any) {
    const response = await apiClient.post(`/bills/${billId}/split/execute`, {
      split_result: splitResult,
      payment_info: paymentInfo,
    });
    return response.data;
  }

  // Get bill participants from blockchain
  static async getBillParticipants(billId: number) {
    const response = await apiClient.get(`/bills/${billId}/participants`);
    return response.data;
  }

  // Get specific participant info by address
  static async getParticipantInfo(billId: number, address: string) {
    const response = await apiClient.get(`/bills/${billId}/participants/${address}`);
    return response.data;
  }

  // Get enhanced bill summary with blockchain data
  static async getBillSummaryWithParticipants(billId: number) {
    const response = await apiClient.get(`/bills/${billId}/summary`);
    return response.data;
  }

  /**
   * Calculate equal split
   */
  static async calculateEqualSplit(request: EqualSplitRequest): Promise<SplitResult> {
    const response = await apiClient.post(`/bills/${request.bill_id}/split/equal`, request);
    return response.data.result;
  }

  /**
   * Calculate custom split
   */
  static async calculateCustomSplit(request: CustomSplitRequest): Promise<SplitResult> {
    const response = await apiClient.post(`/bills/${request.bill_id}/split/custom`, request);
    return response.data.result;
  }

  /**
   * Calculate item-based split
   */
  static async calculateItemSplit(request: ItemSplitRequest): Promise<SplitResult> {
    const response = await apiClient.post(`/bills/${request.bill_id}/split/items`, request);
    return response.data.result;
  }

  /**
   * Validate a split configuration
   */
  static async validateSplit(request: SplitValidationRequest): Promise<SplitValidationResult> {
    const response = await apiClient.post(`/bills/${request.bill_id}/split/validate`, request);
    return response.data.result;
  }
}

// Cache for split options to prevent repeated requests
const splitOptionsCache = new Map<number, { data: SplitOptions; timestamp: number }>();
const CACHE_DURATION = 30000; // 30 seconds

// Request throttling to prevent rapid successive calls
const pendingRequests = new Map<number, Promise<SplitOptions>>();

// Hook for using splitting API with error handling and caching
export const useSplittingAPI = () => {
  const getSplitOptions = async (billId: number) => {
    try {
      // Check cache first
      const cached = splitOptionsCache.get(billId);
      if (cached && Date.now() - cached.timestamp < CACHE_DURATION) {
        return cached.data;
      }

      // Check if there's already a pending request for this billId
      const pending = pendingRequests.get(billId);
      if (pending) {
        return await pending;
      }

      // Create new request and store it as pending
      const requestPromise = SplittingAPI.getSplitOptions(billId);
      pendingRequests.set(billId, requestPromise);

      try {
        const data = await requestPromise;
        
        // Cache the result
        splitOptionsCache.set(billId, {
          data,
          timestamp: Date.now()
        });
        
        return data;
      } finally {
        // Remove from pending requests
        pendingRequests.delete(billId);
      }
    } catch (error: any) {
      // Handle rate limiting specifically
      if (error?.response?.status === 429) {
        console.warn(`Rate limited for bill ${billId}, using cached data if available`);
        const cached = splitOptionsCache.get(billId);
        if (cached) {
          console.log('Returning stale cached data due to rate limiting');
          return cached.data;
        }
      }
      console.error('Failed to get split options:', error);
      throw error;
    }
  };

  const calculateEqualSplit = async (request: EqualSplitRequest) => {
    try {
      return await SplittingAPI.calculateEqualSplit(request);
    } catch (error) {
      console.error('Failed to calculate equal split:', error);
      throw error;
    }
  };

  const calculateCustomSplit = async (request: CustomSplitRequest) => {
    try {
      return await SplittingAPI.calculateCustomSplit(request);
    } catch (error) {
      console.error('Failed to calculate custom split:', error);
      throw error;
    }
  };

  const calculateItemSplit = async (request: ItemSplitRequest) => {
    try {
      return await SplittingAPI.calculateItemSplit(request);
    } catch (error) {
      console.error('Failed to calculate item split:', error);
      throw error;
    }
  };

  const validateSplit = async (request: SplitValidationRequest) => {
    try {
      return await SplittingAPI.validateSplit(request);
    } catch (error) {
      console.error('Failed to validate split:', error);
      throw error;
    }
  };

  const executeSplitPayment = async (billId: number, splitResult: SplitResult, paymentInfo: any) => {
    try {
      return await SplittingAPI.executeSplitPayment(billId, splitResult, paymentInfo);
    } catch (error) {
      console.error('Failed to execute split payment:', error);
      throw error;
    }
  };

  const getBillParticipants = async (billId: number) => {
    try {
      return await SplittingAPI.getBillParticipants(billId);
    } catch (error) {
      console.error('Failed to get bill participants:', error);
      throw error;
    }
  };

  const getParticipantInfo = async (billId: number, address: string) => {
    try {
      return await SplittingAPI.getParticipantInfo(billId, address);
    } catch (error) {
      console.error('Failed to get participant info:', error);
      throw error;
    }
  };

  const getBillSummaryWithParticipants = async (billId: number) => {
    try {
      return await SplittingAPI.getBillSummaryWithParticipants(billId);
    } catch (error) {
      console.error('Failed to get bill summary:', error);
      throw error;
    }
  };

  const clearSplitOptionsCache = (billId?: number) => {
    if (billId) {
      splitOptionsCache.delete(billId);
      pendingRequests.delete(billId);
    } else {
      splitOptionsCache.clear();
      pendingRequests.clear();
    }
  };

  return {
    getSplitOptions,
    calculateEqualSplit,
    calculateCustomSplit,
    calculateItemSplit,
    validateSplit,
    executeSplitPayment,
    getBillParticipants,
    getParticipantInfo,
    getBillSummaryWithParticipants,
    clearSplitOptionsCache,
  };
};

export default SplittingAPI;

File: backend/internal/handlers/splitting.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"payverge/internal/blockchain"
	"payverge/internal/database"
	"payverge/internal/splitting"
)

// SplittingHandler handles bill splitting requests
type SplittingHandler struct {
	db         *database.DB
	splitter   *splitting.SplittingService
	blockchain *blockchain.BlockchainService
}

// NewSplittingHandler creates a new splitting handler
func NewSplittingHandler(db *database.DB, blockchainService *blockchain.BlockchainService) *SplittingHandler {
	return &SplittingHandler{
		db:         db,
		splitter:   splitting.NewSplittingService(db),
		blockchain: blockchainService,
	}
}

// CalculateEqualSplit calculates equal split for a bill
// POST /api/v1/bills/:id/split/equal
func (h *SplittingHandler) CalculateEqualSplit(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	billID, err := strconv.ParseUint(billIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid bill ID"})
		return
	}

	var req splitting.EqualSplitRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Override bill ID from URL parameter
	req.BillID = uint(billID)

	// Validate number of people
	if req.NumPeople <= 0 || req.NumPeople > 20 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Number of people must be between 1 and 20"})
		return
	}

	result, err := h.splitter.CalculateEqualSplit(req.BillID, req.NumPeople)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"result":  result,
	})
}

// CalculateCustomSplit calculates custom split for a bill
// POST /api/v1/bills/:id/split/custom
func (h *SplittingHandler) CalculateCustomSplit(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	billID, err := strconv.ParseUint(billIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid bill ID"})
		return
	}

	var req splitting.CustomSplitRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Override bill ID from URL parameter
	req.BillID = uint(billID)

	// Validate amounts
	if len(req.Amounts) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Amounts cannot be empty"})
		return
	}

	if len(req.Amounts) > 20 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot split between more than 20 people"})
		return
	}

	result, err := h.splitter.CalculateCustomSplit(req.BillID, req.Amounts, req.People)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"result":  result,
	})
}

// CalculateItemSplit calculates item-based split for a bill
// POST /api/v1/bills/:id/split/items
func (h *SplittingHandler) CalculateItemSplit(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	billID, err := strconv.ParseUint(billIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid bill ID"})
		return
	}

	var req splitting.ItemSplitRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Override bill ID from URL parameter
	req.BillID = uint(billID)

	// Validate item selections
	if len(req.ItemSelections) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Item selections cannot be empty"})
		return
	}

	if len(req.ItemSelections) > 20 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot split between more than 20 people"})
		return
	}

	result, err := h.splitter.CalculateItemSplit(req.BillID, req.ItemSelections, req.People)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"result":  result,
	})
}

// GetBillSplitOptions returns available split options for a bill
// GET /api/v1/bills/:id/split/options
func (h *SplittingHandler) GetBillSplitOptions(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	billID, err := strconv.ParseUint(billIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid bill ID"})
		return
	}

	// Get bill from database
	bill, err := h.db.GetBill(uint(billID))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Bill not found"})
		return
	}

	// Parse bill items from JSON
	var billItems []map[string]interface{}
	if bill.Items != "" {
		if err := json.Unmarshal([]byte(bill.Items), &billItems); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to parse bill items"})
			return
		}
	}

	// Calculate remaining amount to be paid
	remainingAmount := bill.TotalAmount - bill.PaidAmount

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"bill": gin.H{
			"id":               bill.ID,
			"bill_number":      bill.BillNumber,
			"total_amount":     bill.TotalAmount,
			"paid_amount":      bill.PaidAmount,
			"remaining_amount": remainingAmount,
			"subtotal":         bill.Subtotal,
			"tax_amount":       bill.TaxAmount,
			"service_fee_amount": bill.ServiceFeeAmount,
			"status":           bill.Status,
		},
		"items": billItems,
		"split_options": gin.H{
			"equal": gin.H{
				"available":    true,
				"description":  "Split the bill equally among all people",
				"min_people":   1,
				"max_people":   20,
			},
			"custom": gin.H{
				"available":    true,
				"description":  "Specify custom amounts for each person",
				"min_people":   1,
				"max_people":   20,
			},
			"items": gin.H{
				"available":    len(billItems) > 0,
				"description":  "Split based on which items each person ordered",
				"min_people":   1,
				"max_people":   20,
				"total_items":  len(billItems),
			},
		},
	})
}

// ValidateSplit validates a split calculation without saving
// POST /api/v1/bills/:id/split/validate
func (h *SplittingHandler) ValidateSplit(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	billID, err := strconv.ParseUint(billIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid bill ID"})
		return
	}

	var req struct {
		Method          string                     `json:"method"`
		NumPeople       int                        `json:"num_people,omitempty"`
		Amounts         map[string]float64         `json:"amounts,omitempty"`
		ItemSelections  map[string][]string        `json:"item_selections,omitempty"`
		People          map[string]string          `json:"people,omitempty"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var result *splitting.SplitResult

	switch req.Method {
	case "equal":
		if req.NumPeople <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Number of people is required for equal split"})
			return
		}
		result, err = h.splitter.CalculateEqualSplit(uint(billID), req.NumPeople)
	case "custom":
		if len(req.Amounts) == 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Amounts are required for custom split"})
			return
		}
		result, err = h.splitter.CalculateCustomSplit(uint(billID), req.Amounts, req.People)
	case "items":
		if len(req.ItemSelections) == 0 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Item selections are required for item split"})
			return
		}
		result, err = h.splitter.CalculateItemSplit(uint(billID), req.ItemSelections, req.People)
	default:
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid split method. Must be 'equal', 'custom', or 'items'"})
		return
	}

	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"valid":   true,
		"result":  result,
	})
}

// GetBillParticipants gets all participants who have paid for a bill from blockchain
// GET /api/v1/bills/:id/participants
func (h *SplittingHandler) GetBillParticipants(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	
	// For now, return empty participants list if blockchain service is not available
	// This allows the frontend to work without blockchain integration
	if h.blockchain == nil {
		c.JSON(http.StatusOK, gin.H{
			"success":      true,
			"participants": []string{},
			"count":        0,
			"message":      "Blockchain service not available - returning empty participants",
		})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	participants, err := h.blockchain.GetBillParticipants(ctx, billIDStr)
	if err != nil {
		// Return empty list instead of error to allow frontend to continue
		c.JSON(http.StatusOK, gin.H{
			"success":      true,
			"participants": []string{},
			"count":        0,
			"error":        err.Error(),
			"message":      "Blockchain error - returning empty participants",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success":      true,
		"participants": participants,
		"count":        len(participants),
	})
}

// GetParticipantInfo gets information about a specific participant
// GET /api/v1/bills/:id/participants/:address
func (h *SplittingHandler) GetParticipantInfo(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	participantAddress := c.Param("address")
	
	if h.blockchain == nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{"error": "Blockchain service not available"})
		return
	}

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	info, err := h.blockchain.GetParticipantInfo(ctx, billIDStr, participantAddress)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"participant": info,
	})
}

// GetBillSummaryWithParticipants gets bill summary including blockchain participant data
// GET /api/v1/bills/:id/summary
func (h *SplittingHandler) GetBillSummaryWithParticipants(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	billID, err := strconv.ParseUint(billIDStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid bill ID"})
		return
	}

	// Get bill from database
	bill, err := h.db.GetBill(uint(billID))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Bill not found"})
		return
	}

	response := gin.H{
		"success": true,
		"bill": gin.H{
			"id":               bill.ID,
			"bill_number":      bill.BillNumber,
			"total_amount":     bill.TotalAmount,
			"paid_amount":      bill.PaidAmount,
			"remaining_amount": bill.TotalAmount - bill.PaidAmount,
			"status":           bill.Status,
		},
	}

	// Add blockchain data if service is available
	if h.blockchain != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()

		// Get blockchain summary
		blockchainSummary, err := h.blockchain.GetBillSummary(ctx, billIDStr)
		if err == nil {
			response["blockchain"] = blockchainSummary
		}

		// Get participants
		participants, err := h.blockchain.GetBillParticipants(ctx, billIDStr)
		if err == nil {
			response["participants"] = participants
			response["participant_count"] = len(participants)
		}
	}

	c.JSON(http.StatusOK, response)
}

// ExecuteSplitPayment coordinates a split payment execution
// POST /api/v1/bills/:id/split/execute
func (h *SplittingHandler) ExecuteSplitPayment(c *gin.Context) {
	billIDStr := c.Param("bill_id")
	
	var req struct {
		SplitResult *splitting.SplitResult `json:"split_result"`
		PaymentInfo map[string]interface{} `json:"payment_info"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if req.SplitResult == nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Split result is required"})
		return
	}

	// Store split execution info (this would typically trigger payment processing)
	c.JSON(http.StatusOK, gin.H{
		"success": true,
		"message": "Split payment execution initiated",
		"bill_id": billIDStr,
		"method":  req.SplitResult.Method,
		"splits":  len(req.SplitResult.Splits),
	})
}
